torch/distributions/gamma.py

```diff
@@ -65,6 +65,9 @@ def rsample(self, sample_shape=torch.Size()):
     def log_prob(self, value):
         if self._validate_args:
             self._validate_sample(value)
+        if not isinstance(value, torch.Tensor) and \
+           (isinstance(value, int) or isinstance(value, float)):
```

Comment:
This could be modified to:
```
if not self._validate_args and isinstance(value, Number):
    value = torch.tensor(value)
```

Reply:
The issue is that some torch operations (e.g. `log`, `expand`) do not support floats. This is something that desired in the future. In the meanwhile, I do not recommend having an implicit conversion in the function. If a user calls the function with a float, which we do not support officially, there will be an explicit error which tells them that they should convert their float to tensors. They can choose to do so if they want. 

Unfortunately, it means it is more difficult to make all distributions (e.g. gamma, half-cauchy) behave consistently for `log_prob`.

Interchanging the two operations is a simple change that will remain appropriate later, even after tensor operators support floats. I do recommend doing this.

With the `not self._validate_args` the behavior of the function suddenly changes between supporting and not supporting floats. I do not recommend doing this either.

Thoughts?

Reply:
I recommended the `not _validate_args`, because when `_validate_args` is True, a check is done to ensure that the value passed is indeed a Tensor. We test whether an appropriate error is thrown in this case in the distribution test suite.

Not adding this condition based on `_validate_args` will cause those tests to fail.

It was why I dismissed my previous review, suggesting that we move this type check outside of `_validate_sample` to the `log_prob` function for all distributions.

cc: @fritzo

Reply:
Well if `_validate_args` only accept tensors, then this is the contract of those objects. If we want to extend support to non-tensor types, those contracts will need to be updated first.